File: ts/board.ts
import {Pawn} from './pawn';
import {Point} from './point';
import {Stack} from './stack';

export class Board {
    private _pawns : Pawn[][];
        
    public board : Pawn[][];
    public turn : number;
    public turnNumber : number;
    public undoStack : Stack<UndoHistory>;
    
    public constructor(public boardSize : number) {}
    
    public initBoard() {
        this.turn = 0;
        this.turnNumber = 0;
        this.undoStack = new Stack<UndoHistory>();
        this._pawns = [[], []];
        
        this.board = new Array(this.boardSize);
        for (let i = 0; i < this.boardSize; i++)
            this.board[i] = new Array(this.boardSize);
            
        for (let y = 0; y < 2; y++)
            for (let x = 0; x < this.boardSize; x++) {
                let pawn = new Pawn({ x, y }, 1);
                this.board[y][x] = pawn;
                this._pawns[1].push(pawn);
            }
                
        for (let y = this.boardSize - 2; y < this.boardSize; y++)
            for (let x = 0; x < this.boardSize; x++) {
                let pawn = new Pawn({ x, y }, 0);
                this.board[y][x] = pawn;
                this._pawns[0].push(pawn);
            }
    }
    
    public checkEnd() {
        if (!this._pawns[1].length) return 0;
        if (!this._pawns[0].length) return 1;
        if (this.board[0].some(p => p && p.color === 0)) return 0;
        if (this.board[this.boardSize - 1].some(p => p && p.color === 1)) return 1;
        return -1;
    }
    
    public nextTurn() {
        this.turn ^= 1;
        this.turnNumber++;
    }
    
    public previousTurn() {
        this.turn ^= 1;
        this.turnNumber--;
    }
    
    public getPawns(color : number) {
        return this._pawns[color];
    }

    public getPossibleMovesOfAPawn(pawn : Pawn) {
        if (!pawn) return [];
        
        let possibleMoves : Point[] = [];
        let deltay = pawn.color ? 1 : -1;

        for (let deltax = -1; deltax <= 1; deltax++) {
            let x = pawn.position.x + deltax;
            let y = pawn.position.y + deltay;

            if (x >= 0 && x < this.boardSize &&
                y >= 0 && y < this.boardSize &&
                (!this.board[y][x] || (deltax !== 0 && this.board[y][x].color !== pawn.color)))
                    possibleMoves.push({ x, y });
        }
        
        return possibleMoves;
    }
    
    public getPossibleMovesOfPawns(color : number) : { pawn : Pawn, point : Point }[] {
        if (this.checkEnd() != -1) return [];
        
        return Array.prototype.concat.apply([], this._pawns[color].map(p => {
            return this.getPossibleMovesOfAPawn(p).map(point => { return { pawn: p, point }; })
        }));
    }
    
    public movePawn(pawn : Pawn, point : Point) {
        let previousPawn = this.getPawn(point);
        this.undoStack.push({
            movedPawn: pawn,
            previousPosition: JSON.parse(JSON.stringify(pawn.position)),
            previousPawn: previousPawn
        });
        
        if (previousPawn) this._pawns[previousPawn.color].splice(this._pawns[previousPawn.color].indexOf(previousPawn), 1);
        
        this.board[pawn.position.y][pawn.position.x] = undefined;
        pawn.position = point;
        this.board[point.y][point.x] = pawn;
        
        this.nextTurn();
    }
    
    public undoMove() {
        let move = this.undoStack.pop();
        if (move.previousPawn) this._pawns[move.previousPawn.color].push(move.previousPawn);
        this.board[move.movedPawn.position.y][move.movedPawn.position.x] = move.previousPawn;        
        move.movedPawn.position = move.previousPosition;
        this.board[move.previousPosition.y][move.previousPosition.x] = move.movedPawn;
        
        this.previousTurn();
    }
    
    public getPawn(position : Point) {
        return this.board[position.y][position.x];
    }

    public pawnDistance(pawn : Pawn) {
        if (pawn.color)
            return pawn.position.y;
        
        return this.boardSize - 1 - pawn.position.y;
    }
}

interface UndoHistory {
    movedPawn: Pawn,
    previousPosition: Point,
    previousPawn: Pawn
}

File: cpp/board.cpp
#include "board.hpp"
#include <iostream>

void Board::init_board() {
    turn = 0;
    turn_number = 0;
    
    _pawns[0] = std::vector<Pawn *>();
    _pawns[1] = std::vector<Pawn *>();
	_moves = std::vector<Move *>();
    
    for (int i = 0; i < board_size; i++) {
        std::vector<Pawn *> row;
        
        for (int j = 0; j < board_size; j++)
            row.push_back(NULL);
            
        board.push_back(row);
    }
    
    for (int y = 0; y < 2; y++)
        for (int x = 0; x < board_size; x++) {
            Pawn * pawn = new Pawn(new Point(x, y), 1);
            board[y][x] = pawn;
            _pawns[1].push_back(pawn);
        }
        
    for (int y = board_size - 2; y < board_size; y++)
        for (int x = 0; x < board_size; x++) {
            Pawn * pawn = new Pawn(new Point(x, y), 0);
            board[y][x] = pawn;
            _pawns[0].push_back(pawn);
        }

	recalculate_moves();
}

int Board::check_end() {
	bool no_pawns = true;
	for (auto pawn : _pawns[1])
		if (!pawn->dead) {
			no_pawns = false;
			break;
		}
	if (no_pawns) return 0;

	no_pawns = true;
	for (auto pawn : _pawns[0])
		if (!pawn->dead) {
			no_pawns = false;
			break;
		}
	if (no_pawns) return 1;

    
    for (int x = 0; x < board_size; x++) {
        Pawn * p_0 = board[0][x];
        Pawn * p_1 = board[board_size - 1][x];
        
        if (p_0 != NULL && p_0->color == 0) return 0;
        if (p_1 != NULL && p_1->color == 1) return 1;
    }
    
    return -1;
}

void Board::next_turn() {
    turn ^= 1;
    turn_number++;
}

void Board::previous_turn() {
    turn ^= 1;
    turn_number--;
}

std::vector<Pawn *> * Board::get_pawns(int color) {
    return &_pawns[color];
}

std::vector<Move *> * Board::get_possible_moves_of_pawns() {
	return &_moves;  
}

void Board::recalculate_moves() {
	for (auto move : _moves)
		delete move;
	_moves.clear();

	if (check_end() != -1) return;

	for (auto it : _pawns[turn]) {
		Pawn * pawn = it;
		if (pawn->dead) continue;

		int deltay = pawn->color ? 1 : -1;

		for (int deltax = -1; deltax <= 1; deltax++) {
			int x = pawn->position->x + deltax;
			int y = pawn->position->y + deltay;

			if (x >= 0 && x < board_size &&
				y >= 0 && y < board_size &&
				(!board[y][x] || (deltax != 0 && board[y][x]->color != pawn->color)))
				_moves.push_back(new Move(pawn, new Point(x, y)));
		}
	}
}

void Board::move_pawn(Pawn * pawn, Point * point) {
    Pawn * previous_pawn = get_pawn(point);
    undo_stack.push(new UndoHistory(
        pawn,
        new Point(pawn->position->x, pawn->position->y),
        previous_pawn
    ));
    
    if (previous_pawn)
		previous_pawn->dead = true;
    
    board[pawn->position->y][pawn->position->x] = NULL;
    pawn->position->x = point->x;
	pawn->position->y = point->y;
    board[point->y][point->x] = pawn;
    
    next_turn();
	recalculate_moves();
}

void Board::undo_move() {
    UndoHistory * history = undo_stack.top();
    undo_stack.pop();

	if (history->previous_pawn)
		history->previous_pawn->dead = false;

	board[history->moved_pawn->position->y][history->moved_pawn->position->x] = history->previous_pawn;
	history->moved_pawn->position->x = history->previous_position->x;
	history->moved_pawn->position->y = history->previous_position->y;
	board[history->previous_position->y][history->previous_position->x] = history->moved_pawn;
    
	previous_turn();
	recalculate_moves();
	delete history;
}

Pawn * Board::get_pawn(Point * point) {
    return board[point->y][point->x];
}